protocol/transfer-precompile.md

```diff
@@ -0,0 +1,83 @@
+# Purpose
+
+Since its inception, the Celo blockchain provides a feature called "[Token
+Duality](https://specs.celo.org/token_duality.html)", which allows the native
+token to be used both in the usual ETH-like way and as an ERC-20 token.
+
+We would like to bring this feature to more OP-Stack based chains and bring the
+Celo L2 closer to vanilla OP-stack by adding the underlying transfer precompile to
+op-geth.
+
+# Summary
+
+Add a precompile to op-geth that allows transferring native tokens.
+This precompile can only be called by a single privileged contract.
+
+# Problem Statement + Context
+
+Many users and contracts have to deal with both ERC20 tokens and native tokens.
+Since they can't be accessed by the same interface, they either have to be
+handled as two different cases or be explicitly wrapped by an ERC-20 token
+wrapping contract.
+Wrapping tokens is not an ideal solution either, since the wrapped token is
+of limited use (e.g. can't be used to pay gas) and wrapping/unwrapping is
+annoying and incurs additional gas charges.
+
+This can be solved by having a contract that provides an ERC-20 interface
+directly to the native token. To implement such a contract, the blockchain must
+provide a way to transfer native tokens on behalf of the transaction sender.
+
+# Proposed Solution
+
+Add one new blockchain parameter `TransferPrecompileAllowedCaller` and a new precompile.
+The parameter defines the address from which the precompile must be called.
+
+Precompile name: `transfer`  
+Precompile address: 0x0000000000000000000000000000000000000101 (directly after the `P256VERIFY` precompile from Fjord)  
+Parameters (abi-encoded): `address from, address to, uint256 value`  
+Gas costs: 9000  
+Result: `value` native tokens will be transferred from `from` to `to`.  
+No return value  
+
+The precompile can fail for the following reasons:
+* `TransferPrecompileAllowedCaller` is not configured for the blockchain
+* It is not called from the contract at the `TransferPrecompileAllowedCaller` address
+* The `from` address has less than `value` native tokens
+* The input is not exactly 96 bytes long
+* The input parameters can't be parsed correctly
+
+Example usage in solidity:
+
+```solidity
+address constant TRANSFER = address(0x0000000000000000000000000000000000000101);
+
+function _transfer(address to, uint256 value) internal {
```

Comment:
What is the design space of possible `TransferPrecompileAllowedCaller` contracts? In practice would it mostly look like the pseudocode here?

Reply:
I feel like the allowed caller should just be a predeploy that is literally this code, perhaps optimized and its part of the design, the way that i read the design right now makes it sound like the allowed caller can be configured to different contracts

Reply:
I would expect the contract to implement the all of ERC20 in a way that passes through the native balance. So in addition to the transfer, you would have other ERC20 functions like

```solidity
  function decimals() external view returns (uint8) {
    return 18;
  }

  function balanceOf(address _owner) public view returns (uint256) {
    return _owner.balance;
  }
```

and potentially other interfaces (it could be a proxy for upgradibility if you consider that safe enough, other token interfaces, etc). I assume it would be hard to find one contract that fits everyone here, and it would need to be configurable in every case to get the name and symbol right.

Having a predeploy with only the a transfer function that can be called from the full ERC20 contract would not solve anything AFAICS. That function would still need to receive the `from` since `msg.sender` can't be used for that due to the additional call. And it would still need to check that the calling contract is the privileged contract.